hirundo/dataset_optimization.py

```diff
@@ -287,23 +276,88 @@ def check_run_by_id(run_id: str, retry=0) -> Generator[dict, None, None]:
+        Args:
+            run_id: The `run_id` produced by a `run_optimization` call
+            stop_on_manual_approval: If True, the function will return an empty DataFrame if the run is awaiting manual approval
 
-        Yields:
-            Each event will be a dict, where:
-            - `"state"` is PENDING, STARTED, RETRY, FAILURE or SUCCESS
-            - `"result"` is a string describing the progress as a percentage for a PENDING state, or the error for a FAILURE state or the results for a SUCCESS state
+        Returns:
+            A pandas DataFrame with the results of the optimization run
+
+        Raises:
+            HirundoError: If the maximum number of retries is reached or if the run fails
+        """
+        logger.debug("Checking run with ID: %s", run_id)
+        with logging_redirect_tqdm():
+            t = tqdm(total=100.0)
+            for iteration in self._check_run_by_id(run_id):
+                if iteration["state"] == "SUCCESS":
+                    t.set_description("Optimization run completed successfully")
+                    t.n = 100.0
+                    t.refresh()
+                    t.close()
+                    return iteration["result"]
+                elif iteration["state"] == "PENDING":
+                    t.set_description("Optimization run queued and not yet started")
+                    t.n = 0.0
+                    t.refresh()
+                elif iteration["state"] == "STARTED":
+                    t.set_description(
+                        "Optimization run in progress. Downloading dataset"
+                    )
+                    t.n = 0.0
+                    t.refresh()
+                elif iteration["state"] is None:
+                    if (
+                        iteration["result"]
+                        and isinstance(iteration["result"], dict)
+                        and iteration["result"]["result"]
+                        and isinstance(iteration["result"]["result"], str)
+                    ):
+                        current_progress_percentage = float(
+                            iteration["result"]["result"].removesuffix("% done")
+                        )
+                        desc = (
+                            "Optimization run completed. Uploading results"
+                            if current_progress_percentage == 100.0
+                            else "Optimization run in progress"
+                        )
+                        t.set_description(desc)
+                        t.n = current_progress_percentage
+                        t.refresh()
+                elif iteration["state"] == "AWAITING MANUAL APPROVAL":
+                    t.set_description("Awaiting manual approval")
+                    t.n = 100.0
+                    t.refresh()
+                    if stop_on_manual_approval:
+                        t.close()
+                        return pd.DataFrame()
```

Comment:
why return an empty df and not `None`?

Reply:
To avoid allowing the function to return `None` or raising any exception. I can work around that by using an `overload` instead. I will do that.